File: bindings/typescript/src/algorithms/traversal.ts
// Graph traversal algorithms as pure functions.
//
// Each takes a Connection first so Graph can delegate in three lines (see
// graph/index.ts). Mirrors bindings/python/src/graphqlite/algorithms/traversal.py.
import type { Connection } from '../connection.ts';
import { extractAlgoArray, parseTraversalResult, type TraversalResult } from './parsing.ts';
import { escapeString } from '../utils.ts';

export type { TraversalResult };

/** Options for {@link bfs} / {@link dfs}. */
export interface TraversalOptions {
  /**
   * Maximum depth to traverse. **Only `maxDepth < 0` means unlimited** —
   * `maxDepth === 0` is passed through as `bfs('x', 0)`, not treated as
   * unlimited (a common mistake). Defaults to `-1`.
   */
  maxDepth?: number;
}

// Shared body: `RETURN <fn>('<escaped start>')` when unlimited, else the
// 2-arg form with the depth interpolated raw (it's a number, not user text).
function traverse(
  conn: Connection,
  fn: 'bfs' | 'dfs',
  startId: string,
  maxDepth: number,
): TraversalResult[] {
  const start = escapeString(startId);
  const query = maxDepth < 0 ? `RETURN ${fn}('${start}')` : `RETURN ${fn}('${start}', ${maxDepth})`;

  const rows = extractAlgoArray(conn.cypher(query).toList());
  const nodes: TraversalResult[] = [];
  for (const row of rows) {
    const parsed = parseTraversalResult(row);
    if (parsed !== null) {
      nodes.push(parsed);
    }
  }
  return nodes;
}

/**
 * Breadth-first search from a starting node — level by level.
 *
 * @returns Rows of `{ user_id, depth, order }` in traversal order.
 */
export function bfs(
  conn: Connection,
  startId: string,
  options: TraversalOptions = {},
): TraversalResult[] {
  return traverse(conn, 'bfs', startId, options.maxDepth ?? -1);
}

/**
 * Depth-first search from a starting node — down each branch before backtracking.
 *
 * @returns Rows of `{ user_id, depth, order }` in traversal order.
 */
export function dfs(
  conn: Connection,
  startId: string,
  options: TraversalOptions = {},
): TraversalResult[] {
  return traverse(conn, 'dfs', startId, options.maxDepth ?? -1);
}

/** Alias for {@link bfs} (matches Python's `breadth_first_search`). */
export const breadthFirstSearch = bfs;

/** Alias for {@link dfs} (matches Python's `depth_first_search`). */
export const depthFirstSearch = dfs;

File: bindings/rust/src/algorithms/traversal.rs
//! Graph traversal algorithm implementations.

use super::parsing::{algo_rows, extract_int, extract_string};
use super::TraversalResult;
use crate::graph::Graph;
use crate::utils::escape_string;
use crate::Result;

impl Graph {
    fn traverse(
        &self,
        func: &str,
        start_id: &str,
        max_depth: Option<i32>,
    ) -> Result<Vec<TraversalResult>> {
        let esc_start = escape_string(start_id);

        let query = match max_depth {
            Some(depth) => format!("RETURN {}('{}', {})", func, esc_start, depth),
            None => format!("RETURN {}('{}')", func, esc_start),
        };

        let result = self.connection().cypher(&query)?;
        let rows = algo_rows(&result);

        let mut nodes = Vec::new();
        for row in rows.iter() {
            if let Some(user_id) = extract_string(row, "user_id") {
                nodes.push(TraversalResult {
                    user_id,
                    depth: extract_int(row, "depth"),
                    order: extract_int(row, "order"),
                });
            }
        }
        Ok(nodes)
    }

    /// Perform breadth-first search traversal from a starting node.
    ///
    /// # Arguments
    ///
    /// * `start_id` - Starting node's id
    /// * `max_depth` - Maximum depth to traverse (None for unlimited)
    pub fn bfs(&self, start_id: &str, max_depth: Option<i32>) -> Result<Vec<TraversalResult>> {
        self.traverse("bfs", start_id, max_depth)
    }

    /// Perform depth-first search traversal from a starting node.
    ///
    /// # Arguments
    ///
    /// * `start_id` - Starting node's id
    /// * `max_depth` - Maximum depth to traverse (None for unlimited)
    pub fn dfs(&self, start_id: &str, max_depth: Option<i32>) -> Result<Vec<TraversalResult>> {
        self.traverse("dfs", start_id, max_depth)
    }
}
